File: .tmp_polymarket_monitor.py
import requests, json, math, os
from datetime import datetime, timezone

NOW = datetime(2026,3,17,10,17,tzinfo=timezone.utc)
BASE='https://gamma-api.polymarket.com/events'
CACHE='/root/.openclaw/workspace/.polymarket_under1_cache.json'

ALLOWED_TAGS = {
    'politics': '政治',
    'sports': '体育',
    'crypto': '加密',
}
FALLBACK_SPORT_TAGS={'nba','nfl','mlb','nhl','soccer','ufc','tennis','golf','cricket','march-madness','champions-league','premier-league','formula-1','f1'}
FALLBACK_CRYPTO_TAGS={'bitcoin','ethereum','solana','xrp','dogecoin','binance','coinbase','kraken','microstrategy'}

session = requests.Session()
session.headers.update({'User-Agent':'Mozilla/5.0 OpenClaw Polymarket Monitor'})

def classify_event(ev):
    tags = ev.get('tags') or []
    slugs = {str(t.get('slug','')).lower() for t in tags}
    labels = {str(t.get('label','')).lower() for t in tags}
    if 'politics' in slugs or 'politics' in labels or 'geopolitics' in slugs:
        return '政治'
    if 'sports' in slugs or 'sports' in labels or slugs & FALLBACK_SPORT_TAGS or labels & FALLBACK_SPORT_TAGS:
        return '体育'
    if 'crypto' in slugs or 'crypto' in labels or slugs & FALLBACK_CRYPTO_TAGS or labels & FALLBACK_CRYPTO_TAGS:
        return '加密'
    return None

def parse_json_field(v, default):
    if isinstance(v, list): return v
    if not v: return default
    try:
        return json.loads(v)
    except Exception:
        return default

def fnum(v, default=0.0):
    try:
        return float(v)
    except Exception:
        return default

ops=[]
offset=0
limit=200
for _ in range(20):
    try:
        r=session.get(BASE, params={'limit':limit,'offset':offset,'closed':'false'}, timeout=30)
        r.raise_for_status()
        batch=r.json()
    except Exception:
        break
    if not isinstance(batch, list) or not batch:
        break
    for ev in batch:
        category=classify_event(ev)
        if not category:
            continue
        for m in ev.get('markets') or []:
            try:
                if not (m.get('active') and not m.get('closed')):
                    continue
                if m.get('acceptingOrders') is False:
                    continue
                end = datetime.fromisoformat(m['endDate'].replace('Z','+00:00'))
                hours=(end-NOW).total_seconds()/3600
                if hours < 2 or hours > 72:
                    continue
                outcomes=parse_json_field(m.get('outcomes'), [])
                prices=[fnum(x) for x in parse_json_field(m.get('outcomePrices'), [])]
                if len(outcomes) != len(prices) or not prices:
                    continue
                idx=max(range(len(prices)), key=lambda i: prices[i])
                prob=prices[idx]
                if prob < 0.80 or prob > 0.98:
                    continue
                spread=fnum(m.get('spread'), 999)
                if spread > 0.008:
                    continue
                volume=fnum(m.get('volumeNum') if m.get('volumeNum') is not None else m.get('volume'))
                liquidity=fnum(m.get('liquidityNum') if m.get('liquidityNum') is not None else m.get('liquidity'))
                if volume < 10000 or liquidity < 10000:
                    continue
                ops.append({
                    'id': str(m.get('id') or m.get('conditionId') or m.get('slug')),
                    'market': m.get('question') or ev.get('title') or '',
                    'category': category,
                    'direction': outcomes[idx],
                    'prob': prob,
                    'spread': spread,
                    'liquidity': liquidity,
                    'volume': volume,
                    'end': end.strftime('%Y-%m-%d %H:%M UTC'),
                })
            except Exception:
                continue
    if len(batch) < limit:
        break
    offset += limit

ops.sort(key=lambda x:(x['end'], -x['prob'], -x['liquidity'], x['market']))

prev={}
if os.path.exists(CACHE):
    try:
        prev=json.load(open(CACHE,'r',encoding='utf-8'))
    except Exception:
        prev={}

new=[]
cur={}
for o in ops:
    key=o['id']
    cur[key]=o
    old=prev.get(key)
    if not old:
        new.append(o)
        continue
    # significant change threshold
    if abs(o['prob']-old.get('prob',-99)) >= 0.02 or abs(o['spread']-old.get('spread',-99)) >= 0.002 or o['direction'] != old.get('direction'):
        new.append(o)

with open(CACHE,'w',encoding='utf-8') as f:
    json.dump(cur,f,ensure_ascii=False,indent=2)

print(json.dumps({'count':len(ops),'new_count':len(new),'items':new}, ensure_ascii=False))

File: tmp_polymarket_monitor.js

async function fetchAll(){
  let all=[];
  for (let offset=0; offset<2500; offset+=500){
    const url='https://gamma-api.polymarket.com/markets?closed=false&active=true&limit=500&offset='+offset;
    const res=await fetch(url);
    const data=await res.json();
    if (!Array.isArray(data) || !data.length) break;
    all=all.concat(data);
    if (data.length<500) break;
  }
  return all;
}
function classify(m){
  const t=((m.question||'')+' '+(m.events||[]).map(e=>e.title||'').join(' ')+' '+(m.events||[]).map(e=>e.slug||'').join(' ')+' '+(m.slug||'')).toLowerCase();
  const sports = /(nba|nfl|mlb|nhl|ncaa|soccer|football|basketball|baseball|tennis|golf|f1|formula 1|ufc|boxing|cricket|ipl|champion|tournament|playoff|world cup|olympic|epl|la liga|serie a|bundesliga|mavericks|bucks|thunder|spurs|natus|sk gaming|lol|esports)/;
  const crypto = /(bitcoin|btc|eth|ethereum|solana|sol|xrp|doge|crypto|token|airdrop|fdv|launch|binance|coinbase|opbnb|memecoin|stablecoin)/;
  const politics = /(trump|president|presidential|prime minister|election|nominee|democrat|republican|senate|house|government|gov shutdown|fed decision|fed |ukraine|russia|iran|ceasefire|netanyahu|hungary|diplomatic|regime|visit china|us forces|dhs shutdown|hormuz|parliamentary)/;
  if (sports.test(t)) return '体育';
  if (crypto.test(t)) return '加密';
  if (politics.test(t)) return '政治';
  return null;
}
function parseJSONMaybe(x){ if (Array.isArray(x)) return x; try{return JSON.parse(x)}catch{return null}}
(async()=>{
  const now = new Date('2026-03-30T13:23:00Z').getTime();
  const markets = await fetchAll();
  const out=[];
  for (const m of markets){
    if (!m.active || m.closed || !m.acceptingOrders) continue;
    const cat=classify(m); if (!cat) continue;
    const end = new Date(m.endDate || m.endDateIso || 0).getTime();
    if (!end || Number.isNaN(end)) continue;
    const hours=(end-now)/36e5;
    if (hours < 2 || hours > 72) continue;
    const liq = Number(m.liquidityNum ?? m.liquidity ?? 0);
    const vol = Number(m.volumeNum ?? m.volume ?? 0);
    if (Math.max(liq,vol) < 10000) continue;
    const spread = Number(m.spread ?? ((m.bestAsk!=null && m.bestBid!=null)? (Number(m.bestAsk)-Number(m.bestBid)) : NaN));
    if (!(spread <= 0.008)) continue;
    const outcomes = parseJSONMaybe(m.outcomes) || [];
    const prices = (parseJSONMaybe(m.outcomePrices) || []).map(Number);
    if (outcomes.length!==2 || prices.length!==2) continue;
    let idx= prices[0]>=prices[1]?0:1;
    const p=prices[idx];
    if (!(p>=0.80 && p<=0.96)) continue;
    out.push({
      id:m.id,
      q:m.question.replace(/s+/g,' ').trim(),
      cat,
      dir:outcomes[idx],
      p,
      spread,
      liq: Math.max(liq,vol),
      end:new Date(end).toISOString().replace('.000Z','Z')
    });
  }
  out.sort((a,b)=>new Date(a.end)-new Date(b.end));
  console.log(JSON.stringify(out,null,2));
})();
